src/rslice.rs

```diff
@@ -0,0 +1,305 @@
+
+/// A read-only view on a Rust byte slice.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API.
+/// `len` indicates the number of bytes than can be safely read.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not dereference the data pointer
+/// beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_bytes<'a> {
+    pub data: *const u8,
+    pub len: size_t,
+    phantom: PhantomData<&'a [u8]>,
+}
+
+impl<'a> From<&'a [u8]> for rustls_slice_bytes<'a> {
+    fn from(s: &[u8]) -> Self {
+        rustls_slice_bytes {
+            data: s.as_ptr(),
+            len: s.len(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[test]
+fn test_rustls_slice_bytes() {
+    let bytes = b"abcd";
+    let rsb: rustls_slice_bytes = bytes.as_ref().into();
+    unsafe {
+        assert_eq!(*rsb.data, b'a');
+        assert_eq!(*rsb.data.offset(3), b'd');
+        assert_eq!(rsb.len, 4);
+    }
+}
+
+/// A read-only view of a slice of Rust byte slices.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API. Because Vec and slice are not `#[repr(C)]`, we
+/// provide access via a pointer to an opaque struct and an accessor method
+/// that acts on that struct to get entries of type `rustls_slice_bytes`.
+/// Internally, the pointee is a `&[&[u8]]`.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not call its methods beyond the
+/// allowed lifetime.
+pub struct rustls_slice_slice_bytes<'a> {
+    pub inner: &'a [&'a [u8]],
+}
+
+/// Return the length of the outer slice. If the input pointer is NULL,
+/// returns 0.
+#[no_mangle]
+pub extern "C" fn rustls_slice_slice_bytes_len(input: *const rustls_slice_slice_bytes) -> size_t {
+    unsafe {
+        match input.as_ref() {
+            Some(c) => c.inner.len(),
+            None => 0,
+        }
+    }
+}
+
+/// Retrieve the nth element from the input slice of slices. If the input
+/// pointer is NULL, or n is greater than the length of the
+/// rustls_slice_slice_bytes, returns rustls_slice_bytes{NULL, 0}.
+#[no_mangle]
+pub extern "C" fn rustls_slice_slice_bytes_get<'a>(
+    input: *const rustls_slice_slice_bytes<'a>,
+    n: size_t,
+) -> rustls_slice_bytes<'a> {
+    let input: &rustls_slice_slice_bytes = unsafe {
+        match input.as_ref() {
+            Some(c) => c,
+            None => {
+                return rustls_slice_bytes {
+                    data: null(),
+                    len: 0,
+                    phantom: PhantomData,
```

Comment:
nit: you could save yourself a little bit of repetition by providing an `std::Default` implementation that constructs a "zero valued" `rustls_slice_bytes`.

Reply:
This is a good point, but I've found on another project that `std::Default` hides things in a way that can cause trouble. I'd like to favor explicitness and obviousness on this project.